src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,423 @@
+ * environment, can be created like this:
+ * {@snippet lang = java :
+ *     Supplier<T> cached = StableValue.newCachedSupplier(original, null);
+ * }
+ * The cached supplier can also be lazily computed by a fresh background thread if a
+ * thread factory is provided as a second parameter as shown here:
+ * {@snippet lang = java :
+ *     Supplier<T> cached = StableValue.newCachedSupplier(original, Thread.ofVirtual().factory());
+ * }
+ * <p>
+ * A cached (also called "memoized") IntFunction, for the allowed given {@code size}
+ * input values {@code [0, size)} and where the given {@code original} IntFunction is
+ * guaranteed to be successfully invoked at most once per inout index even in a
+ * multithreaded environment, can be created like this:
+ * {@snippet lang = java:
+ *     IntFunction<R> cached = StableValue.newCachedIntFunction(size, original, null);
+ *}
+ * Just like a cached supplier, a thread factory can be provided as a second parameter
+ * allowing all the values for the allowed input values to be computed by distinct
+ * background threads.
+ * <p>
+ * A cached (also called "memoized") Function, for the given set of allowed {@code inputs}
+ * and where the given {@code original} function is guaranteed to be successfully invoked
+ * at most once per input value even in a multithreaded environment, can be created like
+ * this:
+ * {@snippet lang = java :
+ *    Function<T, R> cached = StableValue.newCachedFunction(inputs, original, null);
+ * }
+ * Just like a cached supplier, a thread factory can be provided as a second parameter
+ * allowing all the values for the allowed input values to be computed by distinct
+ * background threads.
+ * <p>
+ * A lazy List of stable elements with a given {@code size} and given {@code mapper} can
+ * be created the following way:
+ * {@snippet lang = java :
+ *     List<E> lazyList = StableValue.lazyList(size, mapper);
+ * }
+ * The list can be used to model stable one-dimensional arrays. If two- or more
+ * dimensional arrays are to be modeled, a List of List of ... of E can be used.
+ * <p>
+ * A Map with a given set of {@code keys} and given (@code mapper) associated with
+ * stable values can be created like this:
+ * {@snippet lang = java :
+ *     Map<K, V> lazyMap = StableValue.lazyMap(keys, mapper);
+ * }
+ * <p>
+ * The constructs above are eligible for similar JVM optimizations as StableValue
+ * instances.
+ *
+ * <a id="MemoryConsistency"></a>
+ * <h2>Memory Consistency Properties</h2>
+ * Actions on a presumptive holder value in a thread prior to calling a method that <i>sets</i>
+ * the holder value are seen by any other thread that first <i>observes</i> a set holder value.
+ *
+ * It should be noted that this does <i>not</i> form a proper
+ * <a href="{@docRoot}/java.base/java/util/concurrent/package-summary.html#MemoryVisibility"><i>happens-before</i></a>
+ * relation because the stable value set/observe relation is not transitive.
```

Comment:
The thing is that if thread 1 constructs a value (A) and thread 2 observes the set value (C), then, when we reach C, all the actions up to A are visible.

However, if thread 1 modifies unrelated state (B) then constructs a value (A) then, when we reach C, all the actions up to A are visible but there is no guarantee that B is observed. In other words, there is no transitive guarantee.

Thus, A does not happen before C because B does not happen before C.

Reply:
Can we refer to [JLS's 17.5.1 "Final Field semantics"](https://docs.oracle.com/javase/specs/jls/se22/html/jls-17.html#jls-17.5.1) to describe our stable semantics?